File: 0239-sliding-window-maximum/0239-sliding-window-maximum.cpp
class Solution {
public:
    vector<int> maxSlidingWindow(vector<int>& nums, int k) {
        if(k == 1)
        {
            return nums;
        }
        multiset<int>window;
        for(int i=0;i<k;i++)
        {
            window.insert(nums[i]);
        }
        vector<int>res;
        res.push_back(*window.rbegin());
        int j = 0;
        for(int i=k;i<nums.size();i++)
        {
            window.erase(window.find(nums[j]));
            j++;
            window.insert(nums[i]);
            res.push_back(*window.rbegin());
        }
        return res;
    }
};

File: 0239-sliding-window-maximum/0239-sliding-window-maximum.py
from collections import deque

class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        q = deque()  # stores indices

        for i in range(k):
            while q and nums[q[-1]] <= nums[i]:
                q.pop()
            q.append(i)

        res.append(nums[q[0]])

        for i in range(k, len(nums)):
            while q and q[0] <= i - k:
                q.popleft()

            while q and nums[q[-1]] <= nums[i]:
                q.pop()

            q.append(i)
            res.append(nums[q[0]])

        return res